pytealext/evaluator/evaluator.py

```diff
@@ -693,10 +707,53 @@ def eval_teal(  # pylint: disable=too-many-locals,too-many-branches,too-many-sta
         elif op == "b":
             current_line = branch_targets[args[0]]
         elif op == "callsub":
-            call_stack.append(current_line)
+            call_stack.append(Frame(current_line, len(stack)))
             current_line = branch_targets[args[0]]
         elif op == "retsub":
-            current_line = call_stack.pop()
+            if len(call_stack) == 0:
+                raise Panic("retsub with empty call stack", current_line)
+            frame = call_stack.pop()
+            if frame.clear:
+                expect = frame.height + frame.retc
+                if len(stack) < expect:
+                    raise Panic(
+                        f"retsub with stack size {len(stack)} but expected at least {expect}",
+                        current_line,
+                    )
+                argstart = frame.height - frame.argc
+                bottom_stack = stack[:argstart]
+                returns = stack[frame.height : expect]
+                stack = bottom_stack + returns
+
+            current_line = frame.ret_line
+        elif op == "proto":
+            if prev_op != "callsub":
+                raise Panic("proto must only be used after callsub", current_line)
+            call_stack[-1].argc = int(args[0])
+            call_stack[-1].retc = int(args[1])
+            call_stack[-1].clear = True
+        elif op in ("frame_dig", "frame_bury"):
+            arg_slot = int(args[0])  # should be negative
+            if len(call_stack) == 0:
+                raise Panic("frame_dig with empty call stack", current_line)
+            frame = call_stack[-1]
+            if frame.clear and -arg_slot > frame.argc:
+                raise Panic(
+                    f"frame_dig with arg_slot {arg_slot} but argc {frame.argc}",
+                    current_line,
+                )
+            index = frame.height + arg_slot
+            if index < 0 or index >= len(stack):
+                raise Panic(
+                    f"index {index} out of stack bounds [0,{len(stack)}]",
+                    current_line,
+                )
+            if op == "frame_dig":
+                stack.append(stack[index])
+            else:  # bury
+                if index == len(stack) - 1:
+                    raise Panic("frame_bury with index on top of stack", current_line)
```

Comment:
If I understand correctly `frame_burry` should fail, because the operation would override the frame pointer, right?

Reply:
ok, that's not true as far as I see in the goalgorand. That's would be simply a nonsense, because the value we want to burry is on the top of stack